workflow/controller/scope_test.go

```diff
@@ -0,0 +1,184 @@
+    key: accesskey
+    name: my-minio-cred
+  secretKeySecret:
+    key: secretkey
+    name: my-minio-cred
+`
+
+var ArtifactoryArtifact = `
+name: artifactory-artifact
+path: some/local/path
+artifactory:
+  url: https://artifactory.example.com/some/local/path
+  usernameSecret:
+    key: accesskey
+    name: my-artifactory-cred
+  passwordSecret:
+    key: secretkey
+    name: my-artifactory-cred
+`
+
+var ArtifactoryArtifactWithSubpath = `
+name: artifactory-artifact
+path: some/local/path
+artifactory:
+  url: https://artifactory.example.com/some/local/path/some/subkey
+  usernameSecret:
+    key: accesskey
+    name: my-artifactory-cred
+  passwordSecret:
+    key: secretkey
+    name: my-artifactory-cred
+`
+
+var GCSArtifact = `
+name: s3-artifact
+path: some/local/path
+s3:
+  bucket: test-bucket
+  key: /path/to/some/key
+  serviceAccountKeySecret:
+    key: accesskey
+    name: my-gcs-cred
+`
+
+var GCSArtifactWithSubpath = `
+name: s3-artifact
+path: some/local/path
+s3:
+  bucket: test-bucket
+  key: /path/to/some/key/some/subkey
+  serviceAccountKeySecret:
+    key: accesskey
+    name: my-gcs-cred
+`
+
+var HDFSArtifact = `
+name: hdfs-artifact
+path: some/local/path
+hdfs:
+  addresses:
+  - my-hdfs-namenode-0.my-hdfs-namenode.default.svc.cluster.local:8020
+  - my-hdfs-namenode-1.my-hdfs-namenode.default.svc.cluster.local:8020
+  path: /path/to/some/key
+  hdfsUser: root
+`
+var HDFSArtifactWithSubpath = `
+name: hdfs-artifact
+path: some/local/path
+hdfs:
+  addresses:
+  - my-hdfs-namenode-0.my-hdfs-namenode.default.svc.cluster.local:8020
+  - my-hdfs-namenode-1.my-hdfs-namenode.default.svc.cluster.local:8020
+  path: /path/to/some/key/some/subkey
+  hdfsUser: root
+`
+
+var OSSArtifact = `
+name: oss-artifact
+path: some/local/path
+oss:
+  endpoint: http://oss-cn-hangzhou-zmf.aliyuncs.com
+  bucket: test-bucket-name
+  key: path/to/some/key
+  accessKeySecret:
+    name: my-oss-credentials
+    key: accessKey
+  secretKeySecret:
+    name: my-oss-credentials
+    key: secretKey
+`
+var OSSArtifactWithSubpath = `
+name: oss-artifact
+path: some/local/path
+oss:
+  endpoint: http://oss-cn-hangzhou-zmf.aliyuncs.com
+  bucket: test-bucket-name
+  key: path/to/some/key/some/subkey
+  accessKeySecret:
+    name: my-oss-credentials
+    key: accessKey
+  secretKeySecret:
+    name: my-oss-credentials
+    key: secretKey
+`
+
+func artifactSubPathResolution(t *testing.T, artifactString string, subPathArtifactString string) {
+	scope := wfScope{
+		tmpl:  nil,
+		scope: make(map[string]interface{}),
+	}
+
+	artifact := unmarshalArtifact(artifactString)
+	originalArtifact := artifact.DeepCopy()
+	artifactWithSubPath := unmarshalArtifact(subPathArtifactString)
+
+	scope.addArtifactToScope("steps.test.outputs.artifacts.art", *artifact)
+
+	// Ensure that normal artifact resolution without adding subpaths works
+	resolvedArtifact, err := scope.resolveArtifact("{{steps.test.outputs.artifacts.art}}")
+	assert.Nil(t, err)
```

Comment:
should use NoError

Reply:
Addressed in 80dd46e9d67ad76397c5cd3ffecdb2159c4aff59